### ISSUE TITLE
RUF013 undocumented limitation `x: CustomClass = None` is not caught

### ISSUE THREAD

**Issue by jond01** [astral-sh/ruff#14018]
Seems like the following should be caught, or at least added to the [documented limitations](https://docs.astral.sh/ruff/rules/implicit-optional/#limitations):

```py
from enum import Enum
from typing import Optional


class Letter(Enum):
    """ok."""

    A = "A"


class CatchMe:
    """class."""


def f(
    letter: Letter = None,  # <--- should be caught
    word: Optional[str] = None,
    catch: CatchMe = None,  # <--- should be caught
) -> None:
    """Doc."""
```

* List of keywords you searched for before creating this issue: RUF013
* A minimal code snippet that reproduces the bug:
* The command you invoked: `ruff check file.py --select RUF013`
* The current Ruff settings: see in the snippet.
* The current Ruff version: 0.7.1

**Comment by dhruvmanila** [comment]
> I assume this is "intentional" because Ruff doesn't know to what types `Letter` resolves. E.g. `Letter` could be defined as `type Letter = str | None` in which case the assignment is valid. Not sure if there's another reason for it as well

Yeah, I think this is pretty much the reason. The logic considers all custom type to be similar to `Any` which is the case here. 

For context, the fallback branch tries to resolve the name `Letter` to a qualified name which doesn't exists and thus fallback to using `Unknown` (similar to `Any`).

https://github.com/astral-sh/ruff/blob/4ece8e5c1e674eb0d751fa55b5a72950aa9d20f0/crates/ruff_linter/src/rules/ruff/typing.rs#L116-L119
